### DISCUSSION TITLE
Unable to pull images from a private registry with repository prefix

### DISCUSSION THREAD

**Discussion by carloscarnero** [siderolabs/talos#8891] [General]
I am going through the [quickstart documentation](https://www.talos.dev/v1.7/introduction/quickstart/), trying to bring up a local cluster with Docker, but I'm doing it under an air-gapped environment. I'm using a corporate registry where I have mirrored all the required images (as per the output of `talosctl image default`,) but they require to put those images in repositories with some prefix they gave me. For instance, instead of `example.com:5000/siderolabs/talos:v1.7.4` I must use `example.com:5000/some/prefix/siderolabs/talos:v1.7.4` (note the `some/prefix` in the repository name.) I first tried with:

```console
talosctl cluster create \
  --image=example.com:5000/some/prefix/siderolabs/talos:v1.7.4 \
  --registry-mirror docker.io=http://example.com:5000 \
  --registry-mirror gcr.io=http://example.com:5000 \
  --registry-mirror ghcr.io=http://example.com:5000 \
  --registry-mirror registry.k8s.io=http://example.com:5000
```

which, indeed pulled the Talos image, but it wasn't able to find all the others. Which I believe is the correct outcome, since they live in a prefixed repository (and I was able to confirm by looking at the container logs.) I even tried appending the prefix to the mirrors:

```console
talosctl cluster create \
  --image=example.com:5000/some/prefix/siderolabs/talos:v1.7.4 \
  --registry-mirror docker.io=http://example.com:5000/some/prefix \
  --registry-mirror gcr.io=http://example.com:5000/some/prefix \
  --registry-mirror ghcr.io=http://example.com:5000/some/prefix \
  --registry-mirror registry.k8s.io=http://example.com:5000/some/prefix
```

but that didn't work (I expected that, since my reading of the `--registry-mirror` option is that it should be a plain URL sans path components.)

So, my question is: how can I tell `talosctl` to use a prefix with all the image repositories?

**Comment by smira** [comment]
It depends on your registry mirror/configuration. You might need to use a config patch instead of flags, as they have only limited support for the most simple case.

If the images have a prefix in their name (vs. the registry path prefix), you can only patch all image references to use yours.

If the prefix is part of the registry path (like https://www.talos.dev/v1.7/talos-guides/configuration/pull-through-cache/#using-harbor-as-a-caching-registry), you can use mirrors to redirect.

**Reply by carloscarnero** [reply]
Thanks for your input! I first tried to use a patch config to modify the images, and I almost got them all except the Flannel ones: I can't find a place to override them? Anyway, using the second method, config patching the mirror configuration did the trick!
